### PR TITLE
in_memory_engine: trigger loading pending region with background task

### LINKED ISSUE #17561: In-memory engine: pending region may never load
## Bug Report

Currently, IME only loads a region when a write operation occurs. For read hotspot regions that rarely or never experience writes, IME fails to cache these hotspots promptly, impacting performance.

### What version of TiKV are you using?

2024/09/19

### PR DESCRIPTION
**Pull request by glorv**
### What is changed and how it works?

Issue Number: Close #17561

What's Changed:

```commit-message
Add a new periodical task to trigger loading pending region. We use a
new casual message to let apply thread triggers loading pending region
(with the same mechanism as `prepare_for_region`).
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by glorv** [comment]
@overvenus @SpadeA-Tang PTAL

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -3810,6 +3810,11 @@ where
         term: u64,
         compact_index: u64,
     },
+    // Trigger loading pending region for in_memory_engine,
+    InMemoryEngineLoadRegion {
+        region_id: u64,
+        cb: Box<dyn FnOnce(&Region) + Send + 'static>,
```
Comment (SpadeA-Tang) [inline components/raftstore/src/store/fsm/apply.rs:3816]: I think it's better to call `trigger_load` or something directly.

**Review by SpadeA-Tang — APPROVED** [review]

**Comment by glorv** [comment]
./retest

**Comment by glorv** [comment]
@overvenus  PTAL

File: tests/failpoints/cases/test_region_cache_engine.rs [hunk 2]
```diff
@@ -665,3 +665,26 @@ fn test_pre_load_when_transfer_ledaer() {
     let region_cache_engine = cluster.sim.rl().get_region_cache_engine(2);
     region_cache_engine.region_cached(&r);
```
Comment (overvenus) [inline tests/failpoints/cases/test_region_cache_engine.rs:666]:
```suggestion
    assert!(region_cache_engine.region_cached(&r));
```

File: tests/failpoints/cases/test_region_cache_engine.rs [hunk 3]
```diff
@@ -665,3 +665,26 @@ fn test_pre_load_when_transfer_ledaer() {
...
+        .unwrap();
+
+    let (tx, rx) = unbounded();
+    fail::cfg_callback("on_apply_in_memory_engine_load_region", move || {
+        tx.send(()).unwrap();
+    })
+    .unwrap();
+
+    rx.recv_timeout(Duration::from_secs(2)).unwrap();
+    range_cache_engine.region_cached(&r);
```
Comment (overvenus) [inline tests/failpoints/cases/test_region_cache_engine.rs:689]: The failpoint can be removed by using `eventually`.

```rust
    test_util::eventually(Duration::from_millis(100), Duration::from_secs(5), || {
        range_cache_engine.region_cached(&r)
    });
```

Comment (glorv) [reply tests/failpoints/cases/test_region_cache_engine.rs]: Yes. But this failpoint can ensure that target code path has executed, so I prefer to use this explicit failpoint.

**Review by overvenus — APPROVED** [review]

**Pull request merged**